design/automated-provisioning-design.md

```diff
@@ -0,0 +1,147 @@
+- **Static provisioning** refers to a user manually deploying a Contour + Envoy instance, plus a Gateway custom resource that describes that Contour + Envoy instance. No automated provisioners are involved; the work is performed by a human. It’s up to that human to ensure that the Gateway spec matches the deployed Contour + Envoy’s configuration.
+- **Dynamic provisioning** refers to the process where a **Gateway provisioner** (a controller running in the cluster) reacts to the creation of a Gateway custom resource by deploying a Contour + Envoy instance that matches the Gateway’s spec. In this process, a human defines a request for load balancing infrastructure, via the Gateway custom resource, and the Gateway provisioner is responsible for fulfilling that request. This term also encompasses the ongoing reconciliation loop between the Gateway custom resource and the underlying Contour + Envoy resources, as well as the deletion process.
+
+It’s possible that we’ll decide on different terms for these concepts going forward (e.g. “managed” vs. “unmanaged”), but this document will stick with the ones defined above.
+
+
+## Goals
+- Create an optional dynamic Gateway provisioner, implemented in `projectcontour/contour` as a subcommand of the `contour` binary and deployed as a separate Deployment, in order to have a conformant Gateway API implementation.
+- Deprecate the Contour operator and associated `Contour` CRD, to focus effort on the Gateway provisioner.
+
+## Design
+The Contour Operator, including the `operator.projectcontour.io/Contour` CRD, will be deprecated.
+We will continue to make the minimal updates required to keep it functioning as-is, and ship new releases as we ship Contour releases, until Gateway API reaches beta (which is expected to be soon, i.e this year) and Contour ships its first release supporting the beta API.
+At that point we will stop maintaining the Operator, including the Contour CRD, and put it in archive mode.
+This conservative deprecation schedule should give current Operator users enough time to migrate off of it, along with a viable alternative if using Gateway API (described below).
+
+A new subcommand will be added to the contour binary within the `projectcontour/contour` repository, here called the Gateway provisioner.
+This subcommand will implement dynamic Gateway provisioning functionality, as described previously.
+Initially, a good portion of the existing Contour Operator code will be copied into `projectcontour/contour` to provide the required functionality; this code will be modified as needed to fully support the Gateway use cases and to more closely adhere to existing Contour code style and standards.
+
+Use of the Gateway provisioner is optional; users who are not interested in dynamic Gateway provisioning or the Gateway API in general are not required to use it.
```

Comment:
Maybe something about how once we've archived the operator, this will be the only supported dynamic provisioning model though?